modules/github_integration/lib/open_project/github_integration/notification_handler/issue_comment.rb

```diff
@@ -0,0 +1,98 @@
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+
+module OpenProject::GithubIntegration
+  module NotificationHandler
+    ##
+    # Handles GitHub issue comment notifications.
+    class IssueComment
+      include OpenProject::GithubIntegration::NotificationHandler::Helper
+
+      COMMENT_ACTIONS = %w[
+        created
+        edited
+      ].freeze
+
+      def process(payload)
+        return unless associated_with_pr?(payload)
+
+        github_system_user = User.find_by(id: payload.fetch('open_project_user_id'))
+        work_packages = find_mentioned_work_packages(comment_body(payload), github_system_user)
+        pull_request = find_pull_request(payload)
+        new_work_packages = without_already_referenced(work_packages, pull_request)
+
+        upsert_partial_pull_request(payload, new_work_packages)
+        comment_on_referenced_work_packages(new_work_packages, github_system_user, notes(payload)) if new_work_packages.any?
+      end
+
+      private
+
+      def associated_with_pr?(payload)
+        payload['issue']['pull_request'].present?
+      end
+
+      def find_pull_request(payload)
+        html_url = payload['issue']['pull_request']['html_url']
+        GithubPullRequest.find_by(github_html_url: html_url)
+      end
+
+      def comment_body(payload)
+        payload.fetch('comment').fetch('body')
+      end
+
+      def upsert_partial_pull_request(payload, work_packages)
+        # Sadly, the webhook data for `issue_comment` events does not give us proper PR data (nor githubs PR id).
+        # Thus, we have to search for the only data we have: html_url.
+        # Even worse, when the PR is unknown to us, we don't have any useful data to create a GithubPullRequest record.
+        # We still want to create a PR record (even if it just has partial data), to remember that it was referenced
+        # and avoid adding reference-comments twice.
+        OpenProject::GithubIntegration::Services::UpsertPartialPullRequest.new.call(
+          github_html_url: payload['issue']['pull_request']['html_url'],
+          number: payload['issue']['number'],
+          repository: payload['repository']['full_name'],
+          work_packages: work_packages
+        )
+      end
+
+      # rubocop:disable Metrics/AbcSize
+      def notes(payload)
+        return unless COMMENT_ACTIONS.include?(payload['action'])
+
+        I18n.t("github_integration.pull_request_referenced_comment",
+               pr_number: payload['issue']['number'],
+               pr_title: payload['issue']['title'],
+               pr_url: payload['comment']['html_url'],
+               repository: payload['repository']['full_name'],
+               repository_url: payload['repository']['html_url'],
+               github_user: payload['comment']['user']['login'],
+               github_user_url: payload['comment']['user']['html_url'])
```

Comment:
Given that you start to reference paths twice (e.g. payload['issue']['pull_request']['html_url']) it might make sense to either:
* wrap the payload by a decorator so that one can then call a getter on it: `payload.issue_pr_html_url`
* extract the path referencing into a method `issue_pr_html_url(payload)`

For deeply nested hash paths there is also `dig` as well so you could also call 
`payload.dig('issue', 'pull_request', 'html_url')`
It is robust in case the path does not exist. On the other hand, it might be better to fail hard in case the expected paths do not exist.

Reply:
My intention was to fail hard (to notify us if/when github changes their webhooks apis without us noticing).
So, I chose `.fetch` over `.dig` most of the times. For some reason I didn't do that here. let me try to make this a little cleaner :) 🏗️

Reply:
ah, I remember why I didn't. (and it's a lame excusion), this specific part was copied from the old implementation :)